presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java

```diff
@@ -234,35 +237,58 @@ private void enforceTimeLimits()
         }
     }
 
+    private static class QueryAndTaskCount<T extends TrackedQuery>
+    {
+        private final T query;
+        private final int taskCount;
+
+        public QueryAndTaskCount(T query, int taskCount)
+        {
+            this.query = query;
+            this.taskCount = taskCount;
+        }
+
+        public T getQuery()
+        {
+            return query;
+        }
+
+        public int getTaskCount()
+        {
+            return taskCount;
+        }
+    }
+
     /**
      *  When cluster reaches max tasks limit and also a single query
      *  exceeds a threshold,  kill this query
      */
-    private void enforceTaskLimits()
+    @VisibleForTesting
+    void enforceTaskLimits()
     {
         int totalRunningTaskCount = 0;
-        int highestRunningTaskCount = 0;
-        Optional<T> highestRunningTaskQuery = Optional.empty();
+        Queue<QueryAndTaskCount> taskCountQueue = new PriorityQueue<>(comparingInt(queryAndCount -> -1 * queryAndCount.getTaskCount()));
+
         for (T query : queries.values()) {
             if (query.isDone() || !(query instanceof QueryExecution)) {
                 continue;
             }
             int runningTaskCount = ((QueryExecution) query).getRunningTaskCount();
             totalRunningTaskCount += runningTaskCount;
-            if (runningTaskCount > highestRunningTaskCount) {
-                highestRunningTaskCount = runningTaskCount;
-                highestRunningTaskQuery = Optional.of(query);
+            if (runningTaskCount > maxQueryRunningTaskCount) {
+                taskCountQueue.add(new QueryAndTaskCount(query, runningTaskCount));
             }
         }
 
         runningTaskCount.set(totalRunningTaskCount);
+        int runningTaskCountAfterKills = totalRunningTaskCount;
```

Comment:
Since it's running every second today, thinking if we should reduce the interval since one run will kill multiple queries. CPU wise expecting some increase, not much.

Reply:
There is impact on other parts of this as well - so I feel that we can explore running this more frequently separately.